cpp/src/arrow/compute/row/encode_internal.h

```diff
@@ -111,7 +111,7 @@ class ARROW_EXPORT RowTableEncoder {
                               const std::vector<KeyColumnArray>& cols_in);
 
   // Data initialized once, based on data types of key columns
-  RowTableMetadata row_metadata_;
+  std::shared_ptr<RowTableMetadata> row_metadata_;
```

Comment:
I was suggesting a raw pointer here. `RowTableEncoder` shouldn't be involved in (partly) managing the lifespan of the `RowTableMetadata`.  Someone else, e.g. `RowArray`, who holds `RowTableEncoder` and/or `RowTableImpl` instances should also hold an instance of `RowTableMetadata`, whose lifespan is fully managed by this "someone else".

Reply:
Ah, I understand.
We should uniformly manage the life cycle of `RowTableMetadata` in `RowArray` and `GrouperFastImpl`, and others like `RowTableImpl` and `RowTableEncoder` only need to hold a raw-pointer of `RowTableMetadata`.

Reply:
Yes, this is what I meant :)